conbench/api/compare.py

```diff
@@ -160,53 +287,131 @@ def get(self, compare_ids) -> f.Response:
-        contender = _result_dict_for_compare_api(contender_benchmark_result)
-        comparator = BenchmarkResultComparator(
-            baseline,
-            contender,
-            threshold,
-            threshold_z,
-        )
+class CompareRunsAPI(ApiEndpoint):
+    @staticmethod
+    def _get_all_results_for_a_run(run_id: str) -> List[BenchmarkResult]:
+        """Get all benchmark results for a run. Abort if the run doesn't exist or if
+        there are no results for the run.
+        """
+        result = Session.scalars(
+            s.select(BenchmarkResult).where(BenchmarkResult.run_id == run_id)
+        ).all()
 
-        return comparator.compare()
+        if not result:
+            f.abort(
+                404, description=f"no benchmark results found for run ID: '{run_id}'"
+            )
+
+        return list(result)
+
+    @staticmethod
+    def _join_results(
+        baseline_results: List[BenchmarkResult],
+        contender_results: List[BenchmarkResult],
+    ) -> List[Tuple[Optional[BenchmarkResult], Optional[BenchmarkResult]]]:
+        """Do a full outer join of two lists of benchmark results, pairing by case,
+        context, hardware, repo, and unit.
+
+        If a case/context/hardware/repo/unit combination is present in one list but not
+        the other, it will still be included but the other tuple element will be None.
+
+        If there are multiple results for a case/context/hardware/repo/unit combination
+        in both lists, a cartesian product of them all will be returned.
+        """
+        _key_type = Tuple[str, str, str, Optional[str], Optional[str]]
+
+        def _generate_key(result: BenchmarkResult) -> _key_type:
+            repo = result.run.commit.repository if result.run.commit else None
+            return (
+                result.case_id,
+                result.context_id,
+                result.run.hardware.hash,
+                repo,
+                result.unit,
+            )
 
+        # cchru means 'case/context/hardware/repo/unit'
+        baseline_results_by_cchru: Dict[
+            _key_type, List[Optional[BenchmarkResult]]
+        ] = collections.defaultdict(list)
+        contender_results_by_cchru: Dict[
+            _key_type, List[Optional[BenchmarkResult]]
+        ] = collections.defaultdict(list)
 
-class CompareListEndpoint(ApiEndpoint, CompareMixin):
-    def _get_results(self, id: str) -> List[BenchmarkResult]:
-        raise NotImplementedError
+        for result in baseline_results:
+            baseline_results_by_cchru[_generate_key(result)].append(result)
+
+        for result in contender_results:
+            contender_results_by_cchru[_generate_key(result)].append(result)
+
+        joined_results: List[
+            Tuple[Optional[BenchmarkResult], Optional[BenchmarkResult]]
+        ] = []
+
+        for cchru in set(baseline_results_by_cchru) | set(contender_results_by_cchru):
```

Comment:
thanks for embracing cchru! we should come up with a pronunciation for it! sounds funny for sure!

Reply:
Haha, it's really close to "church". We could backronym that to mean "case/hardware/unit/repository/context/happiness" 🙂